### ISSUE TITLE
Automatic Python Version Trove Classifiers

### ISSUE THREAD

**Issue by harkabeeparolus** [astral-sh/uv#8716]
I wish uv would provide automatic trove classifiers for Python versions when building wheels.

[Poetry adds automatic Python version trove classifiers](https://python-poetry.org/docs/pyproject/#classifiers). This improves the package metadata when uploading to PyPI, and means that the developer doesn't have to manage these versions manually as the project's Python requirements change.

Example (with **Poetry** version 1.8.4):

```bash
$ poetry new foo && cd foo
$ poetry build
$ unzip -p dist/foo-0.1.0-py3-none-any.whl foo-0.1.0.dist-info/METADATA
Metadata-Version: 2.1
Name: foo
Version: 0.1.0
Summary:
Author: Yeeg9yoo Ohl4eida
Author-email: boo6gima@choo7eir.com
Requires-Python: >=3.9,<4.0
Classifier: Programming Language :: Python :: 3
Classifier: Programming Language :: Python :: 3.9
Classifier: Programming Language :: Python :: 3.10
Classifier: Programming Language :: Python :: 3.11
Classifier: Programming Language :: Python :: 3.12
Classifier: Programming Language :: Python :: 3.13
Description-Content-Type: text/markdown
```

Compare this to **uv 0.4.29** (85f9a0d0e 2024-10-30):

```bash
$ uv init --package foo && cd foo
$ uv build
$ unzip -p dist/foo-0.1.0-py3-none-any.whl foo-0.1.0.dist-info/METADATA
Metadata-Version: 2.3
Name: foo
Version: 0.1.0
Summary: Add your description here
Author-email: Yeeg9yoo Ohl4eida <boo6gima@choo7eir.com>
Requires-Python: >=3.13
```

**Comment by harkabeeparolus** [comment]
> Could they automatically be added to the `pyproject.toml` too? (Not sure when but it could be even better)

Preferably not, in my opinion... My point is that adding default classifiers automatically to the wheel enables you to specify versions in _one place_ only, instead of copy-pasting or editing them in several different places.

A _single source of truth_ is better than multiple manually managed sources of confusion. 😉

**Comment by konstin** [comment]
Specification note: PEP 621 only allows us to do this if with `dynamic = ["classifiers"]`. Poetry is allow to do this since they are not using the `[project]` table.

**Comment by HenriBlacksmith** [comment]
> > Could they automatically be added to the `pyproject.toml` too? (Not sure when but it could be even better)
> 
> 
> 
> Preferably not, in my opinion... My point is that adding default classifiers automatically to the wheel enables you to specify versions in _one place_ only, instead of copy-pasting or editing them in several different places.
> 
> 
> 
> A _single source of truth_ is better than multiple manually managed sources of confusion. 😉

Maybe those are two different features then, one could be a command to update "static" classifiers in the `pyproject.toml` and another one a setting to enrich classifiers in the build (maybe via the dynamic mode)

**Comment by harkabeeparolus** [comment]
> Specification note: PEP 621 only allows us to do this if with `dynamic = ["classifiers"]`. Poetry is allowed to do this since they are not using the `[project]` table.

That's a good point!

I'm not familiar enough with PEP 621, so I wonder... Would it be possible to use `dynamic = ["classifiers"]` to get default values, and at the same time also use manually specified values to override the defaults or to specify additional classifiers?

**Comment by harkabeeparolus** [comment]
> Specification note: PEP 621 only allows us to do this if with `dynamic = ["classifiers"]`. Poetry is allowed to do this since they are not using the `[project]` table.

Update:

I browsed through the specifications for [`pyproject.toml`](https://packaging.python.org/en/latest/specifications/pyproject-toml/) as well as [Core metadata](https://packaging.python.org/en/latest/specifications/core-metadata/), which are based partly on PEP 621 and 643, and I see no problems with my suggestion.

While both specifying core metadata AND declaring it to be dynamic is not explicitly mentioned in the specs, it is also clearly not forbidden, as far as I can tell.

In other words, we could certainly have build backends that provide default Python version trove classifiers if unspecified AND if `classifiers` is declared as dynamic... And we could include `dynamic = ["classifiers"]` in the `uv init` templates for the backends that support this feature.

**Comment by tsvikas** [comment]
update:
poetry 2 now works like this:
if `project.classifiers` is set, it doesn't change it
if `project.classifiers` is not set, and `project.dynamic` contains `"classifiers"`: it will set classifiers to be `tool.poetry.classifiers` (if exists) + the automaticly created classifiers (license, version, etc.).

that seems like a reasonable behaviour. could we duplicate it?

**Comment by harkabeeparolus** [comment]
> update: poetry 2 now works like this: if `project.classifiers` is set, it doesn't change it if `project.classifiers` is not set, and `project.dynamic` contains `"classifiers"`: it will set classifiers to be `tool.poetry.classifiers` (if exists) + the automaticly created classifiers (license, version, etc.).
> 
> that seems like a reasonable behaviour. could we duplicate it?

I agree, this would be nice. Poetry 2 seems to do a lot of things right. 😄
More info about this new behavior: https://python-poetry.org/docs/pyproject/#classifiers

**Comment by MicaelJarniac** [comment]
`Typing :: Typed` could also be inferred from the project I think, by the presence of a `py.typed` marker.

**Comment by ambv** [comment]
Note that since this issue was opened, specifying `License ::` in trove classifiers [is now deprecated](https://packaging.python.org/en/latest/specifications/pyproject-toml/#classifiers) and the preferred spelling is to use the license text or file in `[project]`, which lands in `METADATA` without duplication.

**Comment by Avasam** [comment]
> Specification note: PEP 621 only allows us to do this if with `dynamic = ["classifiers"]`. Poetry is allowed to do this since they are not using the `[project]` table.

https://peps.python.org/pep-0808/ should in theory help with this exact scenario, by allowing static metadata that can be expanded (and only expanded) dynamically.

In the mean time, a build backend warning when `Programming Language :: Python :: 3.*` is outside the declared `requires-python` would be nice (and maybe even a warning when missing a *stable released* python version that does fall withing the declared `requires-python` !)
